agent/consul/autopilotevents/ready_servers_events.go

```diff
@@ -0,0 +1,305 @@
+
+func (e EventPayloadReadyServers) HasReadPermission(authz acl.Authorizer) bool {
+	// Any service in the mesh will need access to where the servers live. Therefore
+	// we check if the authorizer grants permissions on any service and if so then
+	// we allow seeing where the servers are.
+	var authzContext acl.AuthorizerContext
+	structs.WildcardEnterpriseMetaInPartition(structs.WildcardSpecifier).
+		FillAuthzContext(&authzContext)
+
+	return authz.ServiceWriteAny(&authzContext) == acl.Allow
+}
+
+func ExtractEventPayload(event stream.Event) (EventPayloadReadyServers, error) {
+	if event.Topic != EventTopicReadyServers {
+		return nil, fmt.Errorf("unexpected topic (%q) for a %q event", event.Topic, EventTopicReadyServers)
+	}
+
+	if payload, ok := event.Payload.(EventPayloadReadyServers); ok {
+		return payload, nil
+	}
+
+	return nil, fmt.Errorf("unexpected payload type %T for %q event", event.Payload, EventTopicReadyServers)
+}
+
+type Config struct {
+	GetStore     func() StateStore
+	Publisher    Publisher
+	timeProvider timeProvider
+}
+
+// ReadyServersEventPublisher is capable to tracking changes to ready servers
+// between consecutive calls to PublishReadyServersEvents. It will then publish
+// "ready-servers" events as necessary.
+type ReadyServersEventPublisher struct {
+	Config
+	previous EventPayloadReadyServers
+
+	snapshotLock sync.RWMutex
+	snapshot     []stream.Event
+}
+
+func NewReadyServersEventPublisher(config Config) *ReadyServersEventPublisher {
+	return &ReadyServersEventPublisher{
+		Config: config,
+		snapshot: []stream.Event{
+			{
+				Topic:   EventTopicReadyServers,
+				Index:   0,
+				Payload: EventPayloadReadyServers{},
+			},
+		},
+	}
+}
+
+//go:generate mockery -name StateStore -inpkg -testonly
+type StateStore interface {
+	GetNodeID(types.NodeID, *acl.EnterpriseMeta) (uint64, *structs.Node, error)
+}
+
+//go:generate mockery -name Publisher -inpkg -testonly
+type Publisher interface {
+	Publish([]stream.Event)
+}
+
+//go:generate mockery -name timeProvider -inpkg -testonly
+type timeProvider interface {
+	Now() time.Time
+}
+
+// PublishReadyServersEvents will publish a "ready-servers" event if the list of
+// ready servers has changed since the last time events were published.
+func (r *ReadyServersEventPublisher) PublishReadyServersEvents(state *autopilot.State) {
+	if events, ok := r.readyServersEvents(state); ok {
+		// if the event publisher were to not be able to keep up with procesing events
+		// then its possible this blocks. It could cause autopilot to not update its
+		// state as often as it should. However if this blocks for over 10s then
+		// not updating the autopilot state as quickly is likely the least of our
+		// concerns. If we need to make this async then we probably need to single
+		// flight these to ensure proper event ordering.
+		r.snapshotLock.Lock()
+		r.snapshot = events
+		r.snapshotLock.Unlock()
+		r.Publisher.Publish(events)
```

Comment:
very tiny nit: it may help readability a bit to put this comment on the `Publish` call rather than the lock-assign-unlock block.

```suggestion
		r.snapshotLock.Lock()
		r.snapshot = events
		r.snapshotLock.Unlock()

		// if the event publisher were to not be able to keep up with procesing events
		// then its possible this blocks. It could cause autopilot to not update its
		// state as often as it should. However if this blocks for over 10s then
		// not updating the autopilot state as quickly is likely the least of our
		// concerns. If we need to make this async then we probably need to single
		// flight these to ensure proper event ordering.
		r.Publisher.Publish(events)
```

Reply:
I moved things around as suggested.